### ISSUE TITLE
Max unconfirmed chainlength 

### ISSUE THREAD

**Issue by Joukehofman** [bitcoin/bitcoin#9752]
### Describe the issue
Because transaction malleability isn't that much of an issue lately, we've decided to let the wallet spend zero conf change. It has happened that we create a transaction just moments before a significant increase in (higher paying) transactions. Because of #9645, there is a chance that you end up with an unconfirmed chain of 24 child transactions. If you let the wallet produce a 25th child, it will return an error because the mempool won't accept it.  This is also problematic in the case we would like to cpfp the whole chain, because other nodes won't relay the transaction to the miners.

#### If so, please list the steps to reproduce below:
1. Predict an increase in transactions. (You need a crystal ball)
2. Create transaction
3. Create more transactions
4. ?
5. Profit

### Expected behaviour
Ideally it would stop if the chain is 24 transactions long, so one could try a cpfp transaction to compensate for the fees.

### Actual behaviour
It will create a transaction chain of 25 transaction, making it hard to have the whole chain confirmed in a certain time.

### What version of bitcoin-core are you using?
Official binary: 0.13.1

**Comment by Joukehofman** [comment]
A quick way would be to have a setting for allowed length of unconfirmed transactions in the mempool, and push the cpfp-transaction on an other node.

**Comment by jonasschnelli** [comment]
I think https://github.com/bitcoin/bitcoin/pull/9262 addresses that issue (partially?).
This fix is available in 0.13.2 (as well as in 0.14).
But #9262 is still disabled by default (`DEFAULT_WALLET_REJECT_LONG_CHAINS = false;`)

@Joukehofman:
You could try to upgrade to 0.13.2 and enable `-walletrejectlongchains=true`.

**Comment by instagibbs** [comment]
Yes `-walletrejectlongchains` sounds like what you want. It will simply reject the proposed transaction, allowing you to try something else.

**Comment by Joukehofman** [comment]
It is indeed what I want! Thanks for mentioning @jonasschnelli and fixing it @instagibbs !

**Comment by instagibbs** [comment]
Your use-case might be an argument for stopping 1 shorter even, to allow for a single CPFP in the mempool(with that one hitting the chain limit). As-is it might not quite work.

**Comment by Joukehofman** [comment]
Ah, I thought that was the intention, but you only catch it if mempool doesn't accept the transaction. But, this option, combined with a setting for allowed mempool length would solve my issue in a nice way.

**Comment by instagibbs** [comment]
Do note that there is no package relay for CPFP, so miners would have had to accepted the transactions into the mempool, but merely considered them too-low of a fee as a package.

**Comment by Joukehofman** [comment]
I know. As far as I can tell the transaction are being propagated on the network as they are seen by our other nodes, it's just that the fees of those transactions start to lag behind the moment an uptrend in transactions is started.

**Comment by morcos** [comment]
@Joukehofman Take a look at -limitancestorcount and -limitdescendantcount (run bitcoind with -help --help-debug).  These will limit all chains your node will accept/relay not just your own transactions, but could serve the purpose you want if you set them to limit your chains to 24 and then submit the 25th tx to the network via a different mechanism.

But if everyone started doing this, network relay would become a mess, so I do think a better solution could be for the wallet to try to stop one transaction shorter if -walletrejectlongchains is true.

**Comment by TheBlueMatt** [comment]
Yea, that also means you have to manually relay the 25th TX. We probably should stop one short by default (or provide an option to do so, somehow I thought we had...)

**Comment by RHavar** [comment]
What's the issue in general with allowing long unconf chains? The risk if one of them is maliated and there's a lot of wasted relay?

 There's certain times it's pretty desirable to create some pretty massive chains, for example a service the does deposits and withdrawals. Deposits go into 1 wallet, withdrawals come out from another wallet. The deposit wallet batch sends money to the withdrawal wallet with an extremely low feerate (e.g. 10 sat/byte) for it to get confirmed when there's not much network activity (e.g. weekends).

But this means the "withdrawal wallet" will be doing hundreds or thousands of sends for every receive it as, so by necessity create a long transaction chain and keep bumping into the unconfirmed chain limit pretty easily.

**Comment by instagibbs** [comment]
@RHavar haven't thought about this a ton, but AFAIK it's the tracking all the chains, which can be quite "wide" in a combinatorial sense, and updating package feerates, etc.

**Comment by TheBlueMatt** [comment]
The issue is the runtime of a bunch of things in the mempool is dependant on chain length, so allowing long chains is a very large potential DoS attack for the CPFP/mempool-size-limiting logic.

**Comment by sipa** [comment]
To be clear, these are reasons why there is a max unconfirmed chainlength _in the mempool_. They're not necessarily reasons for disallowing them in the wallet - as the later parts of the chain can enter the mempool after the first ones get confirmed. However, the Bitcoin Core wallet uses the mempool as a means for checking likelihood of confirmations, so it rejects creation of things that wouldn't enter.

**Comment by morcos** [comment]
@sipa I think by default it does not reject things that wouldn't enter the mempool (although there is an option to do that).  The default behavior is to create them and then periodically try to have them reaccepted to the mempool.

**Comment by pinheadmz** [comment]
adding 2 cents here, if `-walletrejectlongchains=true` were the default, I would saved a lot of time yesterday! https://github.com/bitcoin/bitcoin/issues/10004?_pjax=%23js-repo-pjax-container

**Comment by tamal19** [comment]
tamal
